File: neetcode-150/binary-search/875. Koko Eating Bananas.cs
public class Solution {
    private bool IsPossible(int[] piles, int h, int mid) {
        long totalTime = 0;
        foreach (int p in piles) {
            totalTime += (long)Math.Ceiling((double)p / mid);
        }
        return totalTime <= h;
    }

    public int MinEatingSpeed(int[] piles, int h) {
        int low = 1;
        int high = piles.Max();
        while (low <= high) {
            int mid = low + (high - low) / 2;
            if (IsPossible(piles, h, mid)) {
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }
        return low;
    }
}

File: neetcode-150/binary-search/875. Koko Eating Bananas.cpp
class Solution {
public:
    long long calcHrs(vector<int>& piles, int mid) {
        int n = piles.size();
        long long hrs = 0;
        for (int pile : piles) {
            hrs += ceil((double)pile / mid);
        }
        return hrs;
    }

    int minEatingSpeed(vector<int>& piles, int h) {
        int l = 1, r = *max_element(piles.begin(), piles.end());
        while (l <= r) {
            int mid = (l + r) / 2;
            long long hrs = calcHrs(piles, mid);
            if (hrs <= h) r = mid - 1;
            else l = mid + 1;
        }
        return l;
    }
};